07_Agents/weather_service_agent/00_Function_DIY_Agents.ipynb

```diff
@@ -0,0 +1,1294 @@
+    "\n",
+    "for i in range(5):\n",
+    "    output = invoke_model(next_step).strip()\n",
+    "    done, next_step = single_agent_step(next_step, output)\n",
+    "    if not done:\n",
+    "        display(Pretty(f'{output}'))\n",
+    "    else:\n",
+    "        display(Pretty('Final answer from LLM:\\n'+f'{next_step}'))\n",
+    "        break"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "5a48a0e8-147d-4525-a6b2-68a09af1b2c4",
+   "metadata": {},
+   "source": [
+    "---\n",
+    "## Leveraging Open Source Agents with Bedrock\n",
+    "\n",
+    "As with the previous notebook, there are many ways to orchestrate agent workflows with LLMs. One of the popular open source frameworks for this is [LangChain's agent module](https://python.langchain.com/docs/modules/agents/). LangChain provides a variety of agent types as well as the same [integration to Amazon Bedrock](https://python.langchain.com/docs/integrations/llms/bedrock) being available for these agent based workflows. Let's work on implementing the same workflow as above into the LangChain ecosystem.\n",
+    "\n",
+    "First, we can define our LLM with the `Bedrock` class."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "51045678",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain.llms.bedrock import Bedrock\n",
+    "\n",
+    "agent_llm = Bedrock(\n",
+    "    client=boto3.client(service_name='bedrock-runtime', region_name=os.environ.get(\"AWS_REGION\"),),\n",
+    "    model_id=\"anthropic.claude-instant-v1\",\n",
+    "    model_kwargs={\n",
+    "        \"max_tokens_to_sample\": 500,\n",
+    "        \"temperature\": 0.0,\n",
+    "    },\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d27832b5",
+   "metadata": {},
+   "source": [
+    "Now we will use the same python code as our original implementation, but we will use the `@tool` decorator to specify that this python function is able to be used as tool for LLMs."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "af9f914a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain.tools import tool\n",
+    "import ast\n",
+    "\n",
+    "@tool (\"get_lat_long\")\n",
+    "def get_lat_long(place: str) -> dict:\n",
+    "    \"\"\"Returns the latitude and longitude for a given place name as a dict object of python.\"\"\"\n",
+    "    url = \"https://nominatim.openstreetmap.org/search\"\n",
+    "\n",
+    "    params = {'q': place, 'format': 'json', 'limit': 1}\n",
+    "    response = requests.get(url, params=params).json()\n",
+    "\n",
+    "    if response:\n",
+    "        lat = response[0][\"lat\"]\n",
+    "        lon = response[0][\"lon\"]\n",
+    "        return {\"latitude\": lat, \"longitude\": lon}\n",
+    "    else:\n",
+    "        return None\n",
+    "\n",
+    "@tool (\"get_weather_dict\")\n",
```

Comment:
Can we annotate the previous function or do we need to redefine the function for using with Langchain?

Reply:
we need to annotate separately